src/solver/conv_hip_implicit_gemm_bwd_data_xdlops.cpp

```diff
@@ -0,0 +1,391 @@
+    auto& conv_ptr       = conv_ptrs.at(config.index);
+    auto& data_ctx       = primitive_parameters.CastTo<conv::DataInvokeParams>();
+    const auto& tensors  = data_ctx.tensors;
+    auto argument_ptr    = conv_ptr->MakeArgumentPointer(
+        static_cast<void*>(tensors.out),
+        const_cast<void*>( // NOLINT (cppcoreguidelines-pro-type-const-cast)
+            static_cast<const void*>(tensors.w)),
+        const_cast<void*>( // NOLINT (cppcoreguidelines-pro-type-const-cast)
+            static_cast<const void*>(tensors.in)),
+        args.N,
+        args.K,
+        args.C,
+        args.input,
+        args.filter,
+        args.output,
+        args.strides,
+        args.dilation,
+        args.lPadding,
+        args.rPadding,
+        {},
+        {},
+        {});
+    auto invoker_ptr            = conv_ptr->MakeInvokerPointer();
+    const auto enable_profiling = handle.IsProfilingEnabled();
+
+    float elapsed_time =
+        invoker_ptr->Run(argument_ptr.get(), {handle.GetStream(), enable_profiling});
+    if(enable_profiling)
+    {
+        handle.ResetKernelTime();
+        handle.AccumKernelTime(elapsed_time);
+    }
+}
+#endif
+
+void PerformanceConfigHipImplicitGemmBwdXdlops::HeuristicInit(const ProblemDescription& problem)
+{
+    this->index = 0;
+#if !MIOPEN_BACKEND_HIP || !MIOPEN_USE_COMPOSABLEKERNEL
+    std::ignore = problem;
+#else
+    this->index      = 0;
+    this->total_size = 0;
+    this->kernel_id  = "";
+    switch(problem.conv_problem.GetInDataType())
+    {
+    case miopenHalf: Init<ck::half_t>(problem); break;
+    case miopenFloat: Init<float>(problem); break;
+    case miopenInt8:
+    case miopenInt32:
+    case miopenInt8x4:
+    case miopenBFloat16:
+    case miopenDouble: break;
+    }
+#endif
+}
+
+bool PerformanceConfigHipImplicitGemmBwdXdlops::SetNextValue(const ProblemDescription& problem)
+{
+    if(total_size == -1)
+        this->HeuristicInit(problem);
+    assert(total_size != -1);
+    if((index + 1) < total_size)
+    {
+        ++index;
+        return true;
+    }
+    else
+        return false;
+}
+
+bool PerformanceConfigHipImplicitGemmBwdXdlops::IsValidValue() const { return index < total_size; }
+
+bool PerformanceConfigHipImplicitGemmBwdXdlops::IsValid(const ProblemDescription& problem) const
+{
+#if !MIOPEN_BACKEND_HIP || !MIOPEN_USE_COMPOSABLEKERNEL
+    std::ignore = problem;
+    return false;
+#else
+    switch(problem.conv_problem.GetInDataType())
+    {
+    case miopenHalf: return CheckIsSupportCKArgs<ck::half_t>(problem);
+    case miopenFloat: return CheckIsSupportCKArgs<float>(problem);
+    case miopenInt8:
+    case miopenInt32:
+    case miopenInt8x4:
+    case miopenBFloat16:
+    case miopenDouble: break;
+    }
+    return false;
+#endif
+}
+
+bool PerformanceConfigHipImplicitGemmBwdXdlops::operator==(
+    const PerformanceConfigHipImplicitGemmBwdXdlops& other) const
+{
+    return this->index == other.index;
```

Comment:
@iq136boy Shouldn't other fields of performance config also be compared?

Reply:
`this->kernel_id = other.kernel_id;`
`this->total_size = other.total_size;`

Reply:
What is strange is that the only member of PerformanceConfig that is serialized is `kernel_id`, so it seems like actual performance config data is namely the `kernel_id`. But I do not see how (where) `kernel_id` is used after read from database.

What is used in `GetSolution` (well, actually in `RunCKSolution()` called from the Invoker) is `index`.

Can someone remind me how this works?

Reply:
> @iq136boy Shouldn't other fields of performance config also be compared?

We might need to add the comparison of total_size too. But the kernel_id is not necessary since ck stored kernel_id based on index. If index is the same, kernel_id must be the same.

Reply:
> What is strange is that the only member of PerformanceConfig that is serialized is `kernel_id`, so it seems like actual performance config data is namely the `kernel_id`. But I do not see how (where) `kernel_id` is used after read from database.
> 
> What is used in `GetSolution` (well, actually in `RunCKSolution()` called from the Invoker) is `index`.
> 
> Can someone remind me how this works?

CK stored the kernel_id info based on the index in a vector. The index is used to access the vector to get all the ck info. Then use the info to run/invoke the kernel. One thing I should mentioned is that CK use their own invoker to invoke the kernel.

Reply:
[Important] @iq136boy 
> We might need to add the comparison of total_size too. 

Why?

Reply:
[Informative] @iq136boy 
> One thing I should mentioned is that CK use their own invoker to invoke the kernel.

Note on terminology. 

Invoker is an object (usually, a lambda) produced by InvokerFactory. InvokerFactory is an object (usually, a lambda as well) produced by `GetSolution()`. We often write lambdas right in `GetSolution()`, so InvokerFactory is an outer lambda, while Invoker is an inner one.

Although it is namely `conv_ptr->MakeInvokerPointer()->Run(...)` that calls the kernel, it's only a part of the Invoker.